File: src/_cpp/ml/MenuItemWithText.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "ml/MenuItemWithText.h"
#include "ml/ImageManager.h"
#include "ml/Audio/AudioEngine.h"
using namespace cocos2d;

const std::string kNameText( "text" );
const std::string kNameText2( "text2" );
const std::string kNameImageNormal( "normal" );
const std::string kNameImageSelected( "selected" );
const std::string kNameImageDisabled( "disabled" );

MenuItemImageWithText::MenuItemImageWithText()
: _imageNormal()
, _imageSelected()
, _imageDisabled()
, _font()
, _text()
, _font2()
, _text2()
, _sound()
, _useScaleEffectOnSelected( true )
, _onClick( nullptr )
, _labelNormal()
, _labelSelected()
, _labelDisabled()
, _labelNormal2()
, _labelSelected2()
, _labelDisabled2()

{}

MenuItemImageWithText::~MenuItemImageWithText()
{
	stopAllActions();
	removeAllChildrenWithCleanup( true );
}

MenuItemImageWithText::Pointer MenuItemImageWithText::create(
	const std::string & normalImage,
	const std::string & selectdImage,
	const std::string & disabledImage,
	const std::string & fontBMP,
	const std::string & text,
	const ccMenuCallback& callback )
{
	Pointer ptr = make_intrusive<MenuItemImageWithText>();
	if( ptr && ptr->initWithNormalImage( normalImage, selectdImage, disabledImage, fontBMP, text, callback ) )
		return ptr;
	ptr.reset( nullptr );
	return ptr;
}

MenuItemImageWithText::Pointer MenuItemImageWithText::create(
	const std::string & normalImage,
	const std::string & selectedImage,
	const std::string & fontBMP,
	const std::string & text,
	const ccMenuCallback & callback )
{
	return create( normalImage, selectedImage, normalImage, fontBMP, text, callback );
}

MenuItemImageWithText::Pointer MenuItemImageWithText::create(
	const std::string & normalImage,
	const std::string & selectedImage,
	const ccMenuCallback & callback )
{
	const std::string kEmpty;
	return create( normalImage, selectedImage, normalImage, kEmpty, kEmpty, callback );
}

MenuItemImageWithText::Pointer MenuItemImageWithText::create(
	const std::string & normalImage,
	const ccMenuCallback & callback )
{
	const std::string kEmpty;
	return create( normalImage, normalImage, kEmpty, kEmpty, kEmpty, callback );
}

MenuItemImageWithText::Pointer MenuItemImageWithText::create()
{
	const std::string kEmpty;
	return create( kEmpty, kEmpty, kEmpty, kEmpty, kEmpty, nullptr );
}

bool MenuItemImageWithText::initWithNormalImage(
	const std::string & normalImage,
	const std::string & selectedImage,
	const std::string & disabledImage,
	const std::string & fontBMP,
	const std::string & text,
	const ccMenuCallback& callback )
{
	MenuItemImage::initWithCallback( callback );
	setCallback( callback );
	NodeExt::init();

	setCascadeColorEnabled( true );
	setCascadeOpacityEnabled( true );

	setImageNormal( normalImage );
	setImageSelected( selectedImage );
	setImageDisabled( disabledImage );
	setFont( fontBMP );
	setText( text );

	return true;
}

void MenuItemImageWithText::setCallback( const ccMenuCallback& callback )
{
	auto base = std::bind( &MenuItemImageWithText::on_click, this, std::placeholders::_1 );
	MenuItem::setCallback( base );
	_onClick = callback;
}

void MenuItemImageWithText::setImageNormal( const std::string & imagefile )
{
	if( _imageNormal == imagefile )
		return;
	_imageNormal = imagefile;
	IntrusivePtr<Sprite> image = ImageManager::shared().sprite( _imageNormal );
	MenuItemImage::setNormalImage( image );
	image->setName( kNameImageNormal );
	locateImages();
}

void MenuItemImageWithText::setImageSelected( const std::string & imagefile )
{
	if( _imageSelected == imagefile )
		return;
	_imageSelected = imagefile;
	IntrusivePtr<Sprite> image = ImageManager::shared().sprite( imagefile );
	MenuItemImage::setSelectedImage( image );
	image->setName( kNameImageSelected );
	locateImages();
}

void MenuItemImageWithText::setImageDisabled( const std::string & imagefile )
{
	if( _imageDisabled == imagefile )
		return;
	_imageDisabled = imagefile;
	IntrusivePtr<Sprite> image = ImageManager::shared().sprite( imagefile );
	MenuItemImage::setDisabledImage( image );
	image->setName( kNameImageDisabled );
	locateImages();
}

void MenuItemImageWithText::setText( const std::string & string )
{
	if( _text == string )return;
	_text = string;
	buildText();
}

void MenuItemImageWithText::setFont( const std::string & fontfile )
{
	if( _font == fontfile )return;
	_font = fontfile;
	buildText();
}

void MenuItemImageWithText::buildText()
{
	if( _font.empty() || _text.empty() )
		return;
	auto allocate = [this]( LabelPointer& label, Node * parent )
	{
		assert( parent );

		Point center = Point( parent->getContentSize() / 2 );

		if( !label )
		{
			label.reset( Label::createWithBMFont( _font, _text ) );
			parent->addChild( label );
			label->setName( kNameText );
		}
		else
		{
			label->setBMFontFilePath( _font );
			label->setString( _text );
		}
		label->setPosition( center );
	};

	if( getNormalImage() )allocate( _labelNormal, getNormalImage() );
	if( getSelectedImage() )allocate( _labelSelected, getSelectedImage() );
	if( getDisabledImage() )allocate( _labelDisabled, getDisabledImage() );
}

void MenuItemImageWithText::setText2( const std::string & string )
{
    if( _text2 == string )return;
    _text2 = string;
    buildText2();
}

void MenuItemImageWithText::setFont2( const std::string & fontfile )
{
    if( _font2 == fontfile )return;
    _font2 = fontfile;
    buildText2();
}

void MenuItemImageWithText::buildText2()
{
    if( _font2.empty() || _text2.empty() )
        return;
    auto allocate = [this]( LabelPointer& label2, Node * parent )
    {
        assert( parent );
        
        Point center = Point( parent->getContentSize() / 2 );
        
        if( !label2 )
        {
            auto str = parent->getName();
			label2.reset( Label::createWithBMFont( _font2, _text2 ) );
            parent->addChild( label2 );
            label2->setName( kNameText2 );
            label2->setAnchorPoint(Point(0.5f, 0.5f));
        }
        else
        {
            label2->setBMFontFilePath( _font2 );
            label2->setString( _text2 );
        }
        label2->setPosition( center );
    };
    
    if( getNormalImage() )allocate( _labelNormal2, getNormalImage() );
    if( getSelectedImage() )allocate( _labelSelected2, getSelectedImage() );
    if( getDisabledImage() )allocate( _labelDisabled2, getDisabledImage() );
}

void MenuItemImageWithText::locateImages()
{
	if( !getNormalImage() )
		return;

	Point center = Point( getNormalImage()->getContentSize() / 2 );

	Node * node( nullptr );
	if( (node = getNormalImage()) )
	{
		node->setAnchorPoint( Point( 0.5f, 0.5f ) );
		node->setPosition( center );
		node->setCascadeColorEnabled( true );
		node->setCascadeOpacityEnabled( true );
	}
	if( (node = getSelectedImage()) )
	{
		node->setAnchorPoint( Point( 0.5f, 0.5f ) );
		node->setPosition( center );
		node->setCascadeColorEnabled( true );
		node->setCascadeOpacityEnabled( true );
	}
	if( (node = getDisabledImage()) )
	{
		node->setAnchorPoint( Point( 0.5f, 0.5f ) );
		node->setPosition( center );
		node->setCascadeColorEnabled( true );
		node->setCascadeOpacityEnabled( true );
	}
}

void MenuItemImageWithText::on_click( Ref*sender )
{
	if( _sound.empty() == false )
	{
		AudioEngine::shared().playEffect( _sound, false, 0 );
	}
	if( _onClick )
		_onClick( sender );
}

Rect MenuItemImageWithText::rect() const
{
	//MenuItem::rect();
	Rect result;
	result = Rect( _position.x, _position.y,
				 _contentSize.width, _contentSize.height );

	Node const* node = getNormalImage();
	if( node == nullptr )
		node = this;
	auto size = node->getContentSize();
	auto pos = getPosition();
	auto center = node->getAnchorPoint();
	result.origin = -Point( size.width * center.x, size.height * center.y );
	result.origin += pos;
	result.size = size;

	return result;
}

void MenuItemImageWithText::selected()
{
	MenuItemImage::selected();

	if( _useScaleEffectOnSelected )
	{
		int tag = 0x123;
		auto actionN = EaseIn::create( ScaleTo::create( 0.3f, 0.8f ), 2 );
		auto actionS = actionN->clone();
		actionN->setTag( tag );
		actionS->setTag( tag );
		if( getNormalImage() )
		{
			getNormalImage()->stopActionByTag( tag );
			getNormalImage()->runAction( actionN );
		}
		if( getSelectedImage() )
		{
			getSelectedImage()->stopActionByTag( tag );
			getSelectedImage()->runAction( actionS );
		}
	}
}

void MenuItemImageWithText::unselected()
{
	MenuItemImage::unselected();

	if( _useScaleEffectOnSelected )
	{
		int tag = 0x123;
		auto actionN = EaseIn::create( ScaleTo::create( 0.2f, 1.0f ), 2 );
		auto actionS = actionN->clone();
		actionN->setTag( tag );
		actionS->setTag( tag );
		if( getNormalImage() )
		{
			getNormalImage()->stopActionByTag( tag );
			getNormalImage()->runAction( actionN );
		}
		if( getSelectedImage() )
		{
			getSelectedImage()->stopActionByTag( tag );
			getSelectedImage()->runAction( actionS );
		}
	}
}

void MenuItemImageWithText::setEnabled( bool bEnabled )
{
	if( isEnabled() == bEnabled )
		return;

	MenuItemImage::setEnabled( bEnabled );
	switchAnimation();
}

void MenuItemImageWithText::onEnter()
{
	MenuItemImage::onEnter();
	switchAnimation();
}

void MenuItemImageWithText::onExit()
{
	MenuItemImage::onExit();
	switchAnimation();
}

void MenuItemImageWithText::setSound( const std::string sound )
{
	_sound = sound;
}

void MenuItemImageWithText::switchAnimation()
{
	return;
	//	stopActionByTag( kActionTagMenuItemImageWithText_Enabled );
	//
	//	if( isEnabled() )
	//	{
	//		float part = (CCRANDOM_MINUS1_1() * 0.1f + 1) / 2;
	//		float s = 1.05f + CCRANDOM_0_1() * 0.05f;
	//		float t0 = 1 * part;
	//		float t1 = 2 * part;
	//		float t2 = 1 * part;
	//
	//		auto a0 = ScaleTo::create( t0, 1 * s, 1 / s );
	//		auto a1 = ScaleTo::create( t1, 1 / s, 1 * s );
	//		auto a2 = ScaleTo::create( t2, 1, 1 );
	//		auto action = RepeatForever::create( Sequence::create( a0, a1, a2, nullptr ) );
	//		runAction( action );
	//		action->setTag( kActionTagMenuItemImageWithText_Enabled );
	//	}
}

File: src/_eu/ml/MenuItemWithText.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/
/**TESTED**/
    //TODO: sound on click
var kNameText = "text";
var kNameText2 = "text2";
var kNameImageNormal = "normal";
var kNameImageSelected = "selected";
var kNameImageDisabled = "disabled";


EU.MenuItemImageWithText = cc.MenuItemImage.extend({

    /** Test instance of */
    __MenuItemImageWithText : true,

    _useScaleEffectOnSelected: null,
    _imageNormal: null,
    _imageSelected: null,
    _imageDisabled: null,
    _font: null,
    _text: null,
    _font2: null,
    _text2: null,
    _sound: null,
    _onClick: null,
    _labelNormal: null,
    _labelSelected: null,
    _labelDisabled: null,
    _labelNormal2: null,
    _labelSelected2: null,
    _labelDisabled2: null,

    ctor: function(){
        this._useScaleEffectOnSelected = true;

        cc.MenuItemImage.prototype.ctor.call(this);
        this.initExt();
        if( arguments.length == 3 && cc.isString(arguments[0]) && cc.isFunction(arguments[1] && cc.isObject(arguments[2])) )
            this.initWithNormalImage(arguments[0], "", "", "", "", arguments[1], arguments[2]);
        else if( arguments.length == 4 && cc.isString(arguments[0]) && cc.isString(arguments[1]) && cc.isFunction(arguments[2] && cc.isObject(arguments[3])) )
            this.initWithNormalImage(arguments[0], arguments[1], "", "", "", arguments[2], arguments[3]);
        else if( arguments.length == 7 )
            this.initWithNormalImage( arguments[0], arguments[1], arguments[2], arguments[3], arguments[4], arguments[5], arguments[6]);
        else
            this.initWithNormalImage("", "", "", "", "", null, null);
    },
    /**
     * initialisation with all parameters
     * @param {string} normalImage
     * @param {string} selectedImage
     * @param {string} disabledImage
     * @param {string} fontBMP
     * @param {string} text
     * @param callback
     * @param target
     * @returns {boolean}
     */
    initWithNormalImage: function( normalImage, selectedImage, disabledImage, fontBMP, text, callback, target ){
        callback = callback || function(){};
        /** 1 argument only - for jsb compatibility*/
        this.initWithCallback( callback.bind(target));
        //EU.NodeExt.init();
        this.setCascadeColorEnabled(true);
        this.setCascadeOpacityEnabled(true);
        this.setImageNormal( normalImage );
        this.setImageSelected( selectedImage );
        this.setImageDisabled( disabledImage );
        this.setFont( fontBMP );
        this.setText( text );
        return true;
    },
    /**
     * set callback for activate button
     * @param callback
     * @param target
     */
    setCallback: function( callback, target ) {
        //var base = std::bind( &on_click, this, std::placeholders::_1 );
        // empty function for jsb
        callback = callback || function(){};
        cc.MenuItem.prototype.setCallback.call( this, callback, target );
        //_onClick = callback;
    },
    /**
     * private function
     * listen texture and locateAll images by center
     * @param {string} file
     */
    listenTexture: function( file ){
        /** @type cc.Texture2D */
        var texture = EU.ImageManager.getTextureForKey(file);
        if( texture ) {
            //TODO: recode - there is no isLoaded and addEventListener in jsb
            if (texture.isLoaded == undefined) {
                this.locateImages(texture);
            } else {
                if (texture.isLoaded())
                    this.locateImages(texture);
                else
                    texture.addEventListener("load", this.locateImages, this, texture);
            }
        }
    },
    /**
     * set image for normal state
     * @param {string} file
     */
    setImageNormal: function( file ) {
        if( this._imageNormal == file )
            return;
        this._imageNormal = file;
        var children = this._normalImage ? this._normalImage.children : [];
        var image = EU.ImageManager.sprite( this._imageNormal );
        this.setNormalImage( image );
        this._normalImage.setName( kNameImageNormal );
        this._normalImage._children = children;
        this.listenTexture(file);
    },
    /**
     * set image for selected image
     * @param {string} file
     */
    setImageSelected: function( file ) {
        if( this._imageSelected == file )
            return;
        this._imageSelected = file;
        var children = this._selectedImage ? this._selectedImage.children : [];
        var image = EU.ImageManager.sprite( this._imageSelected );
        this.setSelectedImage( image );
        this._selectedImage.setName( kNameImageSelected );
        this._selectedImage._children = children;
        this.listenTexture(file);
    },
    /**
     * set image for disabled state
     * @param {string} file
     */
    setImageDisabled: function( file ) {
        if( this._imageDisabled == file )
            return;
        this._imageDisabled = file;
        var children = this._disabledImage ? this._disabledImage.children : [];
        var image = EU.ImageManager.sprite( this._imageDisabled );
        this.setDisabledImage( image );
        this._disabledImage.setName( kNameImageDisabled );
        this._disabledImage._children = children;
        this.listenTexture(file);
    },
    /**
     * set text
     * @param {string} string
     */
    setText: function( string ) {
        if( this._text == string )
            return;
        this._text = string;
        this.buildText();
    },
    /**
     * set fnt font file
     * @param {string} fontfile
     */
    setFont: function( fontfile ) {
        if( this._font == fontfile )return;
        this._font = fontfile;
        this.buildText();
    },
    /**
     * private function
     * build all text nodes for every image (normal, selected, disabled)
     */
    buildText: function() {
        if( this._font.length == 0 || this._text.length == 0 )
            return;
        var allocate = function( label, parent, menuitem )
        {
            EU.assert( parent );

            var center = cc.p( parent.getContentSize().width / 2, parent.getContentSize().height / 2 );

            if( !label )
            {
                label = new cc.LabelBMFont(  menuitem._text, menuitem._font, -1, cc.TEXT_ALIGNMENT_LEFT, cc.p(0,0))
                parent.addChild( label );
                label.setName( kNameText );
            }
            else
            {
                label.setFntFile( menuitem._font );
                label.setString( menuitem._text, true );
            }
            label.setPosition( center );
            return label;
        };

        if( this.getNormalImage() )
            this._labelNormal = allocate( this._labelNormal, this.getNormalImage(), this );
        if( this.getSelectedImage() )
            this._labelSelected = allocate( this._labelSelected, this.getSelectedImage(), this );
        if( this.getDisabledImage() )
            this._labelDisabled = allocate( this._labelDisabled, this.getDisabledImage(), this );
    },
    /**
     * ???
     * @param {string} string
     */
    setText2: function( string ) {
        if( this._text2 == string )return;
        this._text2 = string;
        this.buildText2();
    },
    /**
     * ???
     * @param {string} fontfile
     */
    setFont2: function( fontfile ) {
        if( this._font2 == fontfile )return;
        this._font2 = fontfile;
        this.buildText2();
    },
    /**
     * ???
     */
    buildText2: function() {
        if( this._font2.empty() || this._text2.empty() )
            return;
        var allocate = function( label2, parent, menuitem ) {
            EU.assert( parent );

            var center = cc.p( parent.getContentSize().width / 2, parent.getContentSize().height / 2 );

            if( !label2 ) {
                var str = parent.getName();
                label2 = new cc.LabelBMFont(  menuitem._text, menuitem._font, -1, cc.TEXT_ALIGNMENT_LEFT, cc.p(0,0))
                parent.addChild( label2 );
                label2.setName( kNameText2 );
                label2.setAnchorPoint(cc.p(0.5, 0.5));
            }
            else {
                label2.setBMFontFilePath( menuitem_font2 );
                label2.setString( menuitem_text2 );
            }
            label2.setPosition( center );
        };

        if( this.getNormalImage() )allocate( this._labelNormal2, this.getNormalImage(), this );
        if( this.getSelectedImage() )allocate( this._labelSelected2, this.getSelectedImage(), this );
        if( this.getDisabledImage() )allocate( this._labelDisabled2, this.getDisabledImage(), this );
    },
    /**
     * locate all images by center of button
     * @param {cc.Texture2D} texture
     */
    locateImages: function( texture ) {
        if( !this.getNormalImage() )
            return;

        //TODO: recode - there is no isLoaded nor removeEventListener in jsb
        if( texture == undefined || (texture.isLoaded && texture.isLoaded() == false) )
            return;
        if (texture.isLoaded) texture.removeEventListener("load", this );

        var center = cc.p( this.getNormalImage().getContentSize().width / 2, this.getNormalImage().getContentSize().height / 2 );

        var node = null;
        if( (node = this.getNormalImage()) ){
            node.setAnchorPoint( cc.p( 0.5, 0.5 ) );
            node.setPosition( center );
            node.setCascadeColorEnabled( true );
            node.setCascadeOpacityEnabled( true );
        }
        if( (node = this.getSelectedImage()) ) {
            node.setAnchorPoint( cc.p( 0.5, 0.5 ) );
            node.setPosition( center );
            node.setCascadeColorEnabled( true );
            node.setCascadeOpacityEnabled( true );
        }
        if( (node = this.getDisabledImage()) ) {
            node.setAnchorPoint( cc.p( 0.5, 0.5 ) );
            node.setPosition( center );
            node.setCascadeColorEnabled( true );
            node.setCascadeOpacityEnabled( true );
        }
        this.buildText();
    },
    /**
     *
     * @param (cc.Class)sender
     */
    on_click: function( sender ) {
        if( EU.xmlLoader.stringIsEmpty(this._sound) == false ) {
            EU.AudioEngine.playEffect( this._sound, false, 0 );
        }
        if( this._onClick )
            this._onClick( sender );
    },
    /**
     * return touched area of the button
     * @returns {cc.Rect}
     */
    rect: function() {
        //MenuItem::rect();
        var result = cc.rect(this._position.x, this._position.y,
            this._contentSize.width, this._contentSize.height);

        var node = this.getNormalImage();
        if( node == null )
            node = this;
        var size = node.getContentSize();
        var pos = this.getPosition();
        var center = node.getAnchorPoint();
        result.origin = cc.p( -size.width * center.x, -size.height * center.y );
        result.origin += pos;
        result.size = size;

        return result;
    },
    /**
     * run action on touch the button
     */
    selected: function() {
        cc.MenuItem.prototype.selected.call(this);

        if( this._useScaleEffectOnSelected ) {
            cc.ScaleTo
            var tag = 0x123;
            var actionN = cc.scaleTo( 0.3, 0.8).easing(cc.easeIn(2));
            var actionS = actionN.clone();
            actionN.setTag( tag );
            actionS.setTag( tag );
            if( this.getNormalImage() ) {
                this.getNormalImage().stopActionByTag( tag );
                this.getNormalImage().runAction( actionN );
            }
            if( this.getSelectedImage() ) {
                this.getSelectedImage().stopActionByTag( tag );
                this.getSelectedImage().runAction( actionS );
            }
        }
    },
    /**
     * run action on untouch the button
     */
    unselected: function() {
        cc.MenuItem.prototype.unselected.call(this);

        if( this._useScaleEffectOnSelected ) {
            var  tag = 0x123;
            var actionN = cc.scaleTo( 0.2, 1.0 ).easing(cc.easeIn(2));
            var actionS = actionN.clone();
            actionN.setTag( tag );
            actionS.setTag( tag );
            if( this.getNormalImage() ) {
                this.getNormalImage().stopActionByTag( tag );
                this.getNormalImage().runAction( actionN );
            }
            if( this.getSelectedImage() ) {
                this.getSelectedImage().stopActionByTag( tag );
                this.getSelectedImage().runAction( actionS );
            }
        }
    },
    /**
     *
     * @param {bool} bEnabled
     */
    setEnabled: function( bEnabled ) {
        if( this.isEnabled() == bEnabled )
            return;
        cc.MenuItem.prototype.setEnabled.call(this, bEnabled);
    },
    /**
     *
     */
    onEnter: function() {
        //cc.Node.prototype.onEnter.call(this);
        this._super();
    },
    /**
     *
     */
    onExit: function() {
        //cc.Node.prototype.onExit.call(this);
        this._super();
    },
    /**
     * set sound on click
     * @param {string} sound
     */
    setSound: function( sound ) {
        this._sound = sound;
    },
    /**
     * Scale effect button on touch
     * @param value
     */
    useScaleEffect: function(value){
        this._useScaleEffectOnSelected = value;
    }
});
EU.NodeExt.call(EU.MenuItemImageWithText.prototype);

EU.mlMenuItem = EU.MenuItemImageWithText;